scripts/gitlab/hatchet-analysis.py

```diff
@@ -22,32 +23,100 @@
 
 import hatchet as ht
 
-f1 = options.report
-f2 = options.baseline
-
-gf1 = ht.GraphFrame.from_caliperreader(f1)
-gf2 = ht.GraphFrame.from_caliperreader(f2)
-
-gf3 = gf2 - gf1
-
-# Display dataframe columns
-print(gf3.dataframe.columns)
+class GenericFrame(ht.GraphFrame):
+   def __init__(self, gf):
+      generic_dataframe = gf.dataframe.copy()
+      generic_graph = gf.graph.copy()
+      generic_exc_metrics = gf.exc_metrics
+      generic_inc_metrics = gf.inc_metrics
+      generic_default_metric = gf.default_metric  # in newer Hatchet
+      generic_dataframe.iloc[0, generic_dataframe.columns.get_loc('name')] = 'Variant'
+      ii = generic_dataframe.index[0]
+      # fr = ht.frame.Frame({'name': 'Variant', 'type' : 'region'})
+      fr = ht.graphframe.Frame({'name': 'Variant', 'type': 'region'})
+      nn = ht.graphframe.Node(fr)
+      setattr(nn, '_hatchet_nid', ii._hatchet_nid)
+      setattr(nn, '_depth', ii._depth)
+      setattr(nn, 'children', ii.children)
+      generic_dataframe.rename(index={ii: nn}, inplace=True)
+      setattr(generic_graph, 'roots', [nn])
+      super().__init__(generic_graph, generic_dataframe, generic_exc_metrics, generic_inc_metrics)
+
+def ExtractCommonSubtree(gf1: ht.GraphFrame, gf2: ht.GraphFrame, metric: str) -> (ht.GraphFrame):
+   if (gf1.graph == gf2.graph):
+      return gf1
+   else:
+      cc = gf1.deepcopy()
+      cc2 = gf2.deepcopy()
+      cc.unify(cc2)
+      # search for nodes contained in both graphs {0==both, 1==left only, 2==right only}
+      filter_func = lambda x: x["_missing_node"] == 0
+      common_subtree = cc.filter(filter_func, squash=True)
+      # print(common_subtree.dataframe.columns.tolist())
+      # tt is generator object from post order tree traversal, i.e starts down at first set of leaves
+      tt = common_subtree.graph.roots[0].traverse(order="post")
+      s2 = 0.0  # sum accumulated at depth 2
+      s1 = 0.0  # sum accumulated at depth 1
+      s0 = 0.0
+      m3 = sys.float_info.max
+      # replace subtree values with sum of kernels that have run
+      for nn in tt:
+         if nn._depth == 3:
+            if common_subtree.dataframe.loc[nn, metric] < m3:
+               m3 = common_subtree.dataframe.loc[nn, metric]
+         elif nn._depth == 2:
+            s2 = m3
+            s1 += s2
+            common_subtree.dataframe.loc[nn, metric] = s2
+            m3 = sys.float_info.max
```

Comment:
If there is a level 2 node with no child, it looks like this node with get s2 = sys.float_info.max.
I must be missing something.

Reply:
Also, in general there are no node levels > 3, since the tree structure stops at level 3. Technically, however - there could be a case where MPI barrier timing is injected via additional service setup, and possibly see column metrics at level 4, which we currently don't care about. They'll be summed into the tuning at level 3 by Caliper anyways and are usually miniscule in the grand scheme of things.